### DISCUSSION TITLE
Mastodon needs a plan to migrate off its Microsoft/Github repository server

### DISCUSSION THREAD

**Discussion by broukema** [mastodon/mastodon#27765] [General]
### Steps to reproduce the problem

I was trying to find the source of the Mastodon software code at web pages like https://mastodon.social/about, https://mastodon.online/about, https://mstdn.science/about, https://musician.social/about .

### Expected behaviour

I expected to find the free software hosted on a server run by an organisations whose track record is ethically consistent with the aims of community software.

### Actual behaviour

I found that most of the typical running Mastodon instances point to a repository hosted on a web server run by a firm that was bought by a highly unethical corporation called Microsoft.

### Detailed description

It is inconsistent with the goals of having a community based software ecosystem for running community discussions to continue to develop Mastodon software on Github now that Github has been bought by Microsoft, a corporation that is infamous for its ['Embrace, extend and extinguish' policies][1] towards competitors - both proprietary and free-software competitors. 

Proposed Solution
===============
_Individual instances_ can shift to their own forges without waiting. For example, [Framapiaf.org](https://framapiaf.org) points to [a framagit repository](https://framagit.org/framasoft/framapiaf/mastodon).

_Main development forge_
A plan and announcement of when a migration to an ethical repository server (forge) such as [Codeberg][2] is planned should be proposed. This should be _when_, not _if_.

For example, waiting for more maturity for [ForgeFed][3] might be among the criteria. A way of announcing to the community that the 'issues' structure may need to be reconstructed at the new repository server (forge) would be needed. A rough consensus among the community on a migration plan would be a good idea.

While waiting to do the migration, use of a statement similar to [that recommended by the Software Freedom Conservancy][4] would be helpful to make the Mastodon software community's position clear.

* [Related issue][5]

[1]: https://en.wikipedia.org/wiki/Embrace,_extend,_and_extinguish

[2]: https://codeberg.org

[3]: https://forgefed.org

[4]: https://sfconservancy.org/docs/SupportGiveUpGitHub-README-snippet.md - quoted here:

[5]: https://github.com/mastodon/mastodon/issues/7865

````
Note: we encourage you to add the below to your existing `README.md` on your GitHub project.

== We're Using GitHub Under Protest ==

This project is currently hosted on GitHub.  This is not ideal; GitHub is a
proprietary, trade-secret system that is not Free and Open Souce Software
(FOSS).  We are deeply concerned about using a proprietary system like GitHub
to develop our FOSS project.  We have an
[open {bug ticket, mailing list thread, etc.} ](INSERT_LINK) where the
project contributors are actively discussing how we can move away from GitHub
in the long term.  We urge you to read about the
[Give up GitHub](https://GiveUpGitHub.org) campaign from
[the Software Freedom Conservancy](https://sfconservancy.org) to understand
some of the reasons why GitHub is not a good place to host FOSS projects.

If you are a contributor who personally has already quit using GitHub, please
[check this resource](INSERT_LINK) for how to send us contributions without
using GitHub directly.

Any use of this project's code by GitHub Copilot, past or present, is done
without our permission.  We do not consent to GitHub's use of this project's
code in Copilot.

![Logo of the GiveUpGitHub campaign](https://sfconservancy.org/img/GiveUpGitHub.png)
````

### Specifications

Mastodon 4.0.2

**Comment by broukema** [comment]
Here is an excellent classification of forges - there's a whitelist (good to excellent); a greylist; and a blacklist:

https://git.sdf.org/humanacollaborator/humanacollabora/src/branch/master/forge_comparison.md

@Gargron

**Comment by morremeyer** [comment]
I do understand the concern about centralization of software repositories and I share it, but very much disagree with your framing of your concern and the way you put expectations on maintainers to address your specific concern.

Mastodon is AGPL 3.0 licensed. If the use of GitHub bugs you, you can always maintain a mirror or fork somewhere else.

Additionally, the choice of wording by the forge comparison you listed makes it clear that it cares about free software at all cost more than about the actual people using the software by using terms that the community as a whole is moving away from instead of generally accepted terms like „recommendation“ and „not recommended“.

Lastly, this is not a bug.

**Comment by broukema** [comment]
It's not up to the maintainers alone, it should be up to developers, maintainers and the community of sysadmins who manage mastodon instances, and to some degree ordinary users as supporters, to work on together. It would require very little effort from the maintainers to start off a TODO list saying that x, y, z are part of an initial draft plan for the migration.

``"If the use of GitHub bugs you"`` - sorry, but this is not about me personally - please let us not diverge from the main issue. This is a community bug.

Your comment about what is more important: free software itself versus the community using the software doesn't really help find a solution: the only stable long-term solution is a community using free software and developing it on servers running free-software and run by the community rather than by a corporation whose actions are detrimental to that community. (Regarding your comment on terminology - probably you're trying to say that use of the term "blacklist" with negative connotations is a problem related to historical racism; that's a fair comment: you're welcome to post a bug report at https://git.sdf.org/humanacollaborator/humanacollabora/issues rather than just state this here.) It's not just a centralisation problem: Microsoft has an atrocious track record in terms of relations with software and with the software community: this is an uncontroversial statement in the FOSS community.

@morremeyer Maybe you could say in plain English: do you agree or disagree that the main Mastodon development should urgently start a draft plan for migrating off Microsoft/Github?

If you agree, then where is the draft plan?

**Comment by morremeyer** [comment]
> It's not up to the maintainers alone, it should be up to developers, maintainers and the community of sysadmins who manage mastodon instances, and to some degree ordinary users as supporters, to work on together.

I do not disagree. You are a member of this community and bringing up something that in your eyes is an issue. I do not disagree with that, as I pointed out earlier.

> It would require very little effort from the maintainers to start off a TODO list saying that x, y, z are part of an initial draft plan for the migration.

As a maintainer of an Open Source project myself, I do not think it would require „very litte effort“ to plan and execute such a move. It is a big undertaking for a project the size of Mastodon.

> `"If the use of GitHub bugs you"` - sorry, but this is not about me personally - please let us not diverge from the main issue. This is a community bug.

You're right, it's not about you. Sorry for that. However, no matter how one views this issue, in terms of this repository, it is not a bug.

> Maybe you could say in plain English: do you agree or disagree that the main Mastodon development should urgently start a draft plan for migrating off Microsoft/Github?

In plain English: I disagree. 

Why? I agree that hosting (almost) everything on a centralized, commercial repository is dangerous. I disagree with how you're bringing up that issue and the urgency.

You are bringing up something where you have a solution already in mind, in a manner that sounds/reads very demanding to me. The sense I get from your initial post and the comment above is „This is very important, you are endangering your project and Free Software, fix it already!“.

> If you agree, then where is the draft plan?

You are not only raising an issue, you're also proposing a solution for it: Moving to another platform. Given that you know what you'd like the solution to look like, I do consider it reasonable to expect that you contribute to drafting that plan.

**Comment by broukema** [comment]
I said that it would take very little effort ``to start off a TODO list saying that x, y, z are part of an initial draft plan``; I did not say that it would take very little effort to ``implement`` the plan. I agree that it would take a big effort to get the plan to a stage of being viable and accepted by rough consensus. For obvious reasons, rough consensus would be needed among the active developers: the wider community can't force something on the developers if the developers disagree.

To use your summary of my comment, I'm not saying "fix it already!", I'm saying ``"please start a discussion about fixing it so that people can contribute and help develop viable plans!"`` My guess is that discussions on migration have already taken place, but I couldn't find them. See the section

````
We have an
[open {bug ticket, mailing list thread, etc.} ](INSERT_LINK) where the
project contributors are actively discussing how we can move away from GitHub
in the long term.
````

of the Software Freedom Conservancy text I quoted above. This is missing from the current README.md page of this project.

Judging the urgency is difficult, but it's already four years since Microsoft took over, so I don't see any justification for delaying at least an opening of the process.

I'm not proposing to move to a "platform", I'm proposing to move to a server running software, run by a community. The distinction, as I understand it, is that a "platform" typically means a server running unknown software where the server(s) and software are run by an organisation with non-participatory decision-making (e.g. a corporation).

I don't see how there could be any solution apart from moving to a different forge. (I guess hypothetically, a FOSS community organisation could launch a takeover bid for Microsoft or buy Github from Microsoft, or the Github community could launch a crowdfunded appeal to by Github on behalf of ... who? But these are rather farfetched scenarios.)

Anyway, since you're asking for a plan (for simplicity, this message is PD = CC-0, even though I don't really want it to contribute to COPILOT), here's an initial draft.

* Which particular server should be used? 
  * What are the arguments for/against in terms of community, the legal status of the organisation running the server (e.g. citizens' association? small tech firm?), repository software, expected longevity and stability of the server + sysadmin team?
* What specifications of Forgefed would be considered sufficient for the migration? https://codeberg.org/forgefed/forgefed/issues/87
  * What specifications would be considered preferred?
* How would the data of the ``issues`` be migrated?
* How would the community be notified of the various stages of the plan?
* Who would moderate the decision to migrate?
  * What forum (issue on github? irc channel? matrix channel?) would be used to coordinate the decision-making?
* What would be the time scale (e.g. 1 year) before freezing issues (in some sense) on Github?
* What text should be inserted into README.md to alert readers to the opening of this process?

**Comment by roke-julian-lockhart** [comment]
If you were to move, I hope that it would be to gitlab.com or a self-hosted GitLab instance rather than anything else. After using almost every alternative, I realized that not solely were the others no good, they were actually almost all incomparably worse than GitLab. Only GitHub has a few features that it doesn't, but they're minor – stuff like profile `readme.md` files.

I actually prefer GitLab now. Its CI especially is more coherent.

**Comment by markcellus** [comment]
> If you were to move, I hope that it would be to gitlab.com or a self-hosted GitLab instance rather than anything else. 
> 
> I actually prefer GitLab now. Its CI especially is more coherent.


I could be wrong now, but last time I checked, GitLab isn't fully open-sourced. So you never really know what their software is doing with your code (even if it's self-hosted). I'd go with something self-hosted _and_ open sourced, like [gitea](https://gitea.io/en-us/). Or something like [codeberg](https://codeberg.org/) if self-hosting isn't an option.

**Comment by trwnh** [comment]
`source.joinmastodon.org` used to exist but was shut down because of how painful it was to host gitlab. also it ended up being largely unused.

this bit is my personal opinion, but gitlab is kinda confusingly designed compared to github or gitea. unfortunately gitea is still missing a lot of useful features like built-in workflows, a discussion feature, and so on, in addition to the general web UX not being as good. sure, these could be run separately, but that's just additional maintenance burden. however, it *might* make sense to move if the benefits outweigh the costs. say gitea or forgejo or some other server software added federation. that would be something to consider. it would still take a lot of thought, though -- as you would probably still need to separately spin up some ci/cd software as well as some forum software.

**Comment by oifj34f34f** [comment]
Codeberg/self-hosted would be a good option

**Comment by roke-julian-lockhart** [comment]
@trwnh,

> say gitea or forgejo or some other server software added federation

https://github.com/go-gitea/gitea/issues/14186#issue-776082774

**Comment by spirillen** [comment]
I can recommend Gitlab as a self hosted solution with a Free open Source License. This will add so much more features, that any moderator+ account probably never willingly returns to anything else.

**Comment by oifj34f34f** [comment]
> I can recommend Gitlab as a self hosted solution with a Free open Source License. This will add so much more features, that any moderator+ account probably never willingly returns to anything else.

GitLab has a bloated frontend that doesn't work without JS

**Comment by spirillen** [comment]
> GitLab has a bloated frontend that doesn't work without JS

No need to give thumbs down here guys... the statement is 100% true, JS is required.

But on self hosted instance, everything is self served. no third party crap there

**Comment by roke-julian-lockhart** [comment]
> https://github.com/mastodon/mastodon/issues/22572#event-10892675540

Great, @trwnh - now I can't subscribe solely to issue closure. (If you want to respond, tag me in the comment, else I won't see it.)

**Comment by broukema** [comment]
> If you're going to use something else, **do not use Gitea/Codeberg/Forgejo/whatever**. Please. They are not very good, and CI/code checking features aren't very good as well.

software != server != human community != protocol
* [forgejo](https://forgejo.org/releases) is a fork of the forge _software_ `gitea`; while
* _Codeberg_ is a community: https://matrix.to/#/#codeberg-space:matrix.org that runs 
* a server: https://codeberg.org
These are three different things. A fourth thing is the
* [Forgefed](https://forgefed.org) protocol that is working towards allowing a federation of software forges to evolve.

Bugs in the community, on the server, or in the software or the protocol, can all be freely posted and discussed and, possibly, solved.

> I agree it's dangerous to keep everything on one centralized platform, 

It's not a problem with centralisation alone. Microsoft/Github is fundamentally opposed to a sustainable human/software ecosytem -  infrastructure based on opaque, authoritarian decision-making is the opposite of the Fediverse. Ordinary people cannot participate in decision-making on MS/GH's budget plans and decisions on who or what to give money to.

> I think people forget that Mastodon makes uses of tools that use GitHub's API and services. You can't just switch.

Switching will cause some disruption, that's obvious. Switching from the walled-garden social networks to the Fediverse also comes with some disruption and the need for rebuilding human community, software and servers. The #TyrannyOfConvenience is one of the things that Mastodon and the rest of the [Fediverse](https://en.wikipedia.org/wiki/Fediverse) aim to avoid. Sticking to Microsoft/Github would defeat the whole point of community-based, ethical software and servers, especially as [MS/GH has promised to become even more opaque, privacy-violating and insecure](https://en.wikipedia.org/wiki/GitHub_Copilot#Reception).

**Reply by lunaisnotaboy** [reply]
I apologize if you were confused by my wording. I only put it there so people would realize I'm referring to the software _and_ the services.

As for Forgefed, I think it's a terrible idea. I use the fediverse [extensively](https://michiru.me/@luna), so I'm obviously not opposed to it, but I just don't think integrating software forges into the ecosystem is a good idea.

**Reply by lunaisnotaboy** [reply]
And in the future, please reply directly to my post so we can avoid clogging up the discussion (and my inbox).

**Reply by lunaisnotaboy** [reply]
Y'know what? I give up. I wasn't trying to get you to switch, but this is an uphill battle. I lose, you win. Have fun. :v:

**Reply by aschrijver** [reply]
To nuance the distinction a bit further.. Gitea [was forked](https://gitea-open-letter.coding.social/) after choosing a commercial path, into [Forgejo](https://forgejo.org), but it is a 'soft fork' and Forgejo both upstreams and downstreams all contributions. [Codeberg](https://codeberg.org) used to be based on Gitea, but switched to the community-driven Forgejo (and is custodian of the project). Codeberg adds some modifications to Forgejo releases that facilitate its operation at scale.

I use Codeberg as primary forge, moved away from Github, and I am quite happy with it. The UI is very similar, and intuitive with less bells and whistles. Codeberg provides integrated CI (based on Woodpecker, the open fork of Drone.io) and Codeberg Pages to host static websites, same as Github pages. Recently Forgejo/Gitea integrated the codebase that runs Github Actions, and so that part of GH functionality is available as well.

As for forge federation there's multiple developments. Gitea started with exploring implementation of the [ForgeFed](https://forgefed.org) protocol, an extension to ActivityPub. However, the initiative moved to Forgejo who are moving forward with the help of NLnet funding. @lunaisnotaboy need not fear here that this integration does not fit into the Fediverse ecosystem. The effort focuses on connecting forges, so that from your local forge you can collaborate in remote repositories, create issues, PR's and comments as if they are local. So that does not pollute/flood microblogging (Mastodon) timelines in any way.

The other ActivityPub forge-related development takes place at Gitlab, who have added [ActivityPub experimental support](https://docs.gitlab.com/ee/development/activitypub/). Here it is possible to follow project activity on your Mastodon timeline. But the fedizen is in control of what repositories to follow. In a future release ForgeFed support is considered.

**Reply by lunaisnotaboy** [reply]
@aschrijver Please don't tag me in this, I'd rather not fight a losing battle.

**Comment by broukema** [comment]
Rough consensus should be based on [[[WP:NOTVOTE]]s](https://en.wikipedia.org/wiki/WP:NOTVOTE), but for anyone interested, there's a poll at https://github.com/mastodon/mastodon/discussions/18756 .
